katsdptelstate/rdb_writer.py

```diff
@@ -16,145 +16,146 @@
-            Number of keys written to the file.
-        keys_failed : int
-            Number of keys that failed to be written.
-        """
-        if keys is None:
-            self.logger.warning("No keys specified - dumping entire database")
-            keys = self._r.keys(b'*')
-
-        with open(filename, 'wb') as f:
-            f.write(RDB_HEADER)
-            keys_written = 0
-            keys_failed = 0
-            for key in keys:
-                try:
-                    enc_str = self.encode_item(key)
-                except (ValueError, KeyError) as e:
-                    keys_failed += 1
-                    self.logger.error("Failed to save key: {}".format(e))
-                    continue
-                f.write(enc_str)
-                keys_written += 1
-            for dump in supplemental_dumps:
-                f.write(dump)
+    - String length up to 63 - single byte, 6 LSB encode number directly
+    - String length from 64 to 16383 - two bytes, 6 LSB of byte 1 + 8 from byte 2 encode number
+
+    Thereafter the value, encoding according to its appropriate schema is appended.
+    As a shortcut, the Redis DUMP command is used to generate the encoded value string.
+
+    Note: Redis provides a mechanism for optional key expiry, which we ignore here.
+    """
+    try:
+        key_len = encode_len(len(key))
+    except ValueError as e:
+        raise ValueError('Failed to encode key length: {}'.format(e))
+    # The DUMPed value includes a leading type descriptor,
+    # the encoded value itself (including length specifier),
+    # a trailing version specifier (2 bytes) and finally an 8 byte checksum.
+    # The version specified and checksum are discarded.
+    key_type = key_dump[:1]
+    encoded_val = key_dump[1:-10]
+    return key_type + key_len + key + encoded_val
+
+
+@contextmanager
+def _rdb_file(filename):
+    """Create new RDB file and ensure that header and trailer are in place."""
+    with open(filename, 'wb') as f:
+        f.write(RDB_HEADER)
+        try:
+            yield f
+        finally:
             f.write(RDB_TERMINATOR + RDB_CHECKSUM)
-        if not keys_written:
-            self.logger.error("No valid keys found - removing empty file")
-            os.remove(filename)
-        return (keys_written, keys_failed)
 
-    def encode_supplemental_keys(self, client, keys):
-        """Helper class to encode a specific set of keys on a non-prime Redis like client.
-        """
-        return [self._encode_item(key, client.dump(key)) for key in keys]
 
-    def encode_item(self, key):
-        return self._encode_item(key, self._r.dump(key))
+class RDBWriter(object):
+    """RDB file resource that stores keys from one (or more) Redis DBs.
 
-    def _encode_item(self, key, key_dump):
-        """Returns a binary string containing an RDB encoded key and value.
+    Parameters
+    ----------
+    filename : str
+        Destination filename. Will be opened in 'wb'.
```

Comment:
Should it allow file-like objects too?

Reply:
I thought about it a lot, but ended up postponing it. The problem is the `os.remove` in the closing phase that requires a filename. The workarounds / special cases made me drop the idea.